language/functional_tests/src/checker/directives.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    common::{strip, Sp},
+    errors::*,
+};
+use std::iter::Peekable;
+
+enum Token {
+    String(String),
+    QuotedString(String),
+    Whitespace(String),
+}
+
+fn next_token<I>(input: &mut Peekable<I>) -> Result<Option<(Token)>>
+where
+    I: Iterator<Item = (usize, char)>,
+{
+    macro_rules! next {
+        () => {
+            input.next().map(|(_, c)| c)
+        };
+    }
+
+    macro_rules! peek {
+        () => {
+            input.peek().map(|(_, c)| c)
+        };
+    }
+
+    let res = match next!() {
+        Some('"') => {
+            let mut buffer = String::new();
+            loop {
+                match next!() {
+                    Some('"') => break Some(Token::QuotedString(buffer)),
+                    Some('\\') => match next!() {
+                        Some('\\') => buffer.push('\\'),
+                        Some('n') => buffer.push('\n'),
+                        Some('t') => buffer.push('\t'),
+                        Some('r') => buffer.push('\r'),
+                        Some('"') => buffer.push('"'),
+                        Some(c) => bail!("unrecognized escape character \\{}", c),
+                        None => bail!("unclosed escape character"),
+                    },
+                    Some(c) => buffer.push(c),
+                    None => bail!("unclosed string literal"),
+                }
+            }
+        }
+        Some(c) if c.is_ascii_whitespace() => {
+            let mut buffer = String::new();
+            buffer.push(c);
+            loop {
```

Comment:
I guess it might not be that much cleaner... but you could do 
```
while let Some(c) = peek!() {
  if !c.is_whitespace() break; 
  ...
}
```

Reply:
No actually, in case the next character is not a whitespace, it should not be consumed.